# Data statistics

```{r load_data_stats}
load("data/data.Rdata")
```

## Sequencing reads statistics

```{r reads_stats}
sample_preprocessing %>% 
    summarise(Total=sum(reads_post_filt * 150 / 1000000000) %>% round(2), 
              mean=mean(reads_post_filt * 150 / 1000000000) %>% round(2),
              sd=sd(reads_post_filt * 150 / 1000000000) %>% round(2)) %>%
    unite("Average",mean, sd, sep = " Â± ", remove = TRUE) %>%
    tt()
```

## DNA fractions
```{r dna_fractions_stats}
sequence_fractions <- read_counts %>%
  pivot_longer(-genome, names_to = "sample", values_to = "value") %>%
  group_by(sample) %>%
  summarise(mags = sum(value)) %>%
	left_join(sample_preprocessing, by = join_by(sample == sample)) %>%
	select(sample,mags,bases_pre_filt,bases_post_filt,host_bases,metagenomic_bases) %>%
	mutate(mags_bases = mags*150) %>%
	mutate(lowqual_bases = bases_pre_filt - bases_post_filt) %>%
	mutate(unmapped_bases = metagenomic_bases - mags_bases) %>%
	mutate(unmapped_bases = ifelse(unmapped_bases < 0, 0, unmapped_bases)) %>%
	select(sample, lowqual_bases, host_bases, unmapped_bases, mags_bases)

sequence_fractions %>%
  mutate_at(vars(-sample), ~./1000000000) %>%
  rename("Sample"=1, "Low quality"=2, "Mapped to host"=3, "Unmapped"=4, "Mapped to MAGs"=5) %>%
  tt()
```


```{r dna_fractions_plot, message=FALSE, warning=FALSE, fig.height=6, fig.width=10, fig.fullwidth=TRUE}
sequence_fractions %>%
	pivot_longer(!sample, names_to = "fraction", values_to = "value") %>%
	mutate(value = value / 1000000000) %>%
	mutate(fraction = factor(fraction, levels = c("lowqual_bases","host_bases","unmapped_bases","mags_bases"))) %>%
	ggplot(., aes(x = sample, y = value, fill=fraction)) +
	    geom_bar(position="stack", stat = "identity") +
      scale_fill_manual(name="Sequence type",
                    breaks=c("lowqual_bases","host_bases","unmapped_bases","mags_bases"),
                    labels=c("Low quality","Mapped to host","Unmapped","Mapped to MAGs"),
                    values=c("#CCCCCC", "#bcdee1", "#d8b8a3","#93655c"))+
	    labs(x = "Samples", y = "Amount of data (GB)") +
	    theme_classic() +
	    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=6),legend.position = "bottom")
```
```{r samples_host_samples, comment="", echo=FALSE, message=FALSE, warning=FALSE}
left_join(sample_preprocessing, sample_metadata, by=join_by(sample==sample)) %>%
  mutate(host_perc=host_reads/reads_pre_filt*100)  %>% 
  group_by(species) %>% 
  summarise(
    mean_host_perc = mean(host_perc, na.rm = TRUE),
    sd_host_perc = sd(host_perc, na.rm = TRUE),
    max_host_perc = max(host_perc, na.rm = TRUE),
    min_host_perc = min(host_perc, na.rm = TRUE)
  ) %>%
  tt()
```
