# (PART) METAGENOMICS {-}
# Prepare data

## Load data

Load the original data files outputted by the bioinformatic pipeline.

### Sample metadata

```{r load_sample_metadata2, warning=FALSE, comments="", message=FALSE}
sample_metadata <- read_tsv("data/sample_metadata.tsv")
```

### Read counts

```{r load_read_counts, warning=FALSE, comments="", message=FALSE}
read_counts <- read_tsv("data/metagenomics/read_counts.tsv")
```

### Genome coverage

```{r load_genome_hits, warning=FALSE, comments="", message=FALSE}
genome_coverage <- read_tsv("data/metagenomics/genome_coverage.tsv")
```

### Genome taxonomy

```{r load_genome_taxonomy, warning=FALSE, comments="", message=FALSE}
genome_taxonomy <- read_tsv("data/metagenomics/genome_taxonomy.tsv") %>%
    select(user_genome,classification) %>%
      separate(classification, c("domain","phylum","class","order","family","genus","species"),  sep =";") %>%
      rename(genome=1) %>%
      mutate(genome = str_remove(genome, "\\.fa$"))
```

### Genome quality

```{r load_genome_length, warning=FALSE, comments="", message=FALSE}
genome_quality <- read_tsv("data/metagenomics/genome_quality.tsv")
```

### Genome metadata

```{r generate_genome_metadata, warning=FALSE, comments="", message=FALSE}
genome_metadata <- inner_join(genome_taxonomy,genome_quality,by=join_by(genome==genome)) %>%
    select(-c(lineage,binner)) %>%
    rename(length=size)
```

### Genome tree

```{r load_genome_tree, warning=FALSE, comments="", message=FALSE}
genome_tree <- read_tree("data/metagenomics/genome_tree.tre")
genome_tree$tip.label <- str_replace_all(genome_tree$tip.label,"'", "") #remove single quotes in MAG names
genome_tree$tip.label <- str_remove(genome_tree$tip.label, "\\.fa$") #remove .fa suffix
genome_tree <- keep.tip(genome_tree, tip=genome_taxonomy$genome) # keep only MAG tips
```

### Genome annotations

```{r load_genome_annotations, warning=FALSE, comments="", message=FALSE}
genome_annotations <- read_tsv("data/metagenomics/genome_annotations.tsv.xz") %>%
    rename(gene=1, genome=2, contig=3)
```

### Preprocessing statistics

```{r load_preprocessing_statistics, warning=FALSE, comments="", message=FALSE}
sample_preprocessing <- read_tsv("data/metagenomics/preprocessing_stats.tsv") %>%
  mutate(host_bases=host_reads*300)
```

## Create working objects

Transform the original data files into working objects for downstream analyses.

### Filter reads by coverage

```{r filter_coverage, warning=FALSE, comments="", message=FALSE}
min_coverage=0.3
read_counts_filt <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]])) 
```

### Transform reads into genome counts

```{r calculate_genome_counts_unfiltered, warning=FALSE, comments="", message=FALSE}
readlength=150
genome_counts <- read_counts %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

```{r calculate_genome_counts_filtered, warning=FALSE, comments="", message=FALSE}
readlength=150
genome_counts_filt <- read_counts_filt %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

### Distill annotations into GIFTs 

```{r distill_annotations, warning=FALSE, comments="", message=FALSE}
genome_gifts <- distill(genome_annotations,GIFT_db,genomecol=2,annotcol=c(9,10,19), verbosity=F)
```

## Prepare color scheme

[AlberdiLab](www.alberdilab.dk) projects use unified color schemes developed for the [Earth Hologenome Initiative](www.earthhologenome.org), to facilitate figure interpretation.

```{r get_ehi_colors2, warning=FALSE, comments="", message=FALSE}
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    pull(colors, name=phylum)
```

## Wrap working objects

All working objects are wrapped into a single Rdata object to facilitate downstream usage.

```{r wrap_working_objects, warning=FALSE, comments="", message=FALSE}
save(sample_metadata, 
     sample_preprocessing,
     genome_metadata, 
     read_counts, 
     genome_counts, 
     genome_counts_filt, 
     genome_tree,
     genome_gifts,
     phylum_colors,
     file = "data/metagenomics/data.Rdata")
```
